resolver.go

```diff
@@ -344,172 +336,55 @@ func (r *Resolver) LookupAnswersForType(name string, rrType uint16) (answers []d
-        labels, ok := dns.IsDomainName(node.Value)
-
-        if (ok && labels > 0) {
-            rr = &dns.PTR{header, dns.Fqdn(node.Value)}
-        } else {
-            err = &NodeConversionError{
-                Node: node,
-                Message: fmt.Sprintf("Value '%s' isn't a valid domain name", node.Value),
-                AttemptedType: dns.TypePTR}
-        }
-        return
-    },
-
-    dns.TypeSRV: func (node *etcd.Node, header dns.RR_Header) (rr dns.RR, err error) {
-        parts := strings.SplitN(node.Value, "\t", 4)
-
-        if len(parts) != 4 {
-            err = &NodeConversionError{
-                Node: node,
-                Message: fmt.Sprintf("Value %s isn't valid for SRV", node.Value),
-                AttemptedType: dns.TypeSRV}
-        } else {
-
-            priority, err := strconv.ParseUint(parts[0], 10, 16)
-            if err != nil {
-                return nil, err
-            }
-
-            weight, err := strconv.ParseUint(parts[1], 10, 16)
-            if err != nil {
-                return nil, err
-            }
-
-            port, err := strconv.ParseUint(parts[2], 10, 16)
-            if err != nil {
-                return nil, err
-            }
-
-            target := dns.Fqdn(parts[3])
-
-            rr = &dns.SRV{
-                header,
-                uint16(priority),
-                uint16(weight),
-                uint16(port),
-                target}
-        }
-        return
-    },
-
-    dns.TypeSOA: func (node *etcd.Node, header dns.RR_Header) (rr dns.RR, err error) {
-        parts := strings.SplitN(node.Value, "\t", 6)
-
-        if len(parts) < 6 {
-            err = &NodeConversionError{
-                Node: node,
-                Message: fmt.Sprintf("Value %s isn't valid for SOA", node.Value),
-                AttemptedType: dns.TypeSOA}
-        } else {
-            refresh, err := strconv.ParseUint(parts[2], 10, 32)
-            if err != nil {
-                return nil, err
-            }
-
-            retry, err := strconv.ParseUint(parts[3], 10, 32)
-            if err != nil {
-                return nil, err
-            }
+// RRSetExists returns true if RRs exist for the given name and type (value independent)
+func (r *Resolver) RRSetExists(name string, rrType uint16) (exists bool, err error) {
+    answers, err := r.LookupAnswersForType(dns.Fqdn(name), rrType)
+    if err != nil {
+        return false, err
+    }
 
-            expire, err := strconv.ParseUint(parts[4], 10, 32)
-            if err != nil {
-                return nil, err
-            }
+    return len(answers) > 0, nil
+}
 
-            minttl, err := strconv.ParseUint(parts[5], 10, 32)
-            if err != nil {
-                return nil, err
+// RRSetMatches checks that the set of records in the DNS for the given name
+// and type *exactly* match the given RRs: their data must match, and there must
+// be no more or less RRs
```

Comment:
[ Quoting <notifications@github.com> in "Re: [discodns] Add support for dyna..." ]
>>
>> -            minttl, err := strconv.ParseUint(parts[5], 10, 32)
>> -            if err != nil {
>> -                return nil, err
>> +// RRSetMatches checks that the set of records in the DNS for the given name
>> +// and type *exactly* match the given RRs: their data must match, and there must
>> +// be no more or less RRs
>
>The best readable source code confirmation I can find is in [Bundy/bind-10](https://github.com/bundy-dns/bundy/blob/master/src/lib/python/bundy/ddns/session.py#L396-420).
>
>This seems to imply that, given an RRset exists, then the rdata values already in the DNS need to match _exactly and fully_ the rdata values in the prereq (unless I'm misreading that method), which implies that my implementation here is is correct. (specifically, the check on [line 419](https://github.com/bundy-dns/bundy/blob/97ac840fb1f61f2840d9eba91b64a1f5ed938e80/src/lib/python/bundy/ddns/session.py#L419) indicates that no non-matching rdatas can remain after checking for the prereq ones).
>
>[Their tests](https://github.com/bundy-dns/bundy/blob/97ac840fb1f61f2840d9eba91b64a1f5ed938e80/src/lib/python/bundy/ddns/tests/session_tests.py) seem to confirm this (although it's hard to see without reading and mapping out their entire test harness/test data) -- the method should only return true when the prereq contains the exact three rdata values as already existing, not when just one or two are asked for.
>
>I could really do with a second pair of eyes on this :)

Although it is important to get things right, I think these features of dyn.
updates we not used widely. I.e. most use is to send update to the remote
server, without really caring what's already in the zone.

But of course I'm making this claim without any data what so ever :)